### PR TITLE
Refactor send event

### PR DESCRIPTION
**Pull request by CourtneyJC**
Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request updates [CHANGES.rst](https://github.com/matrix-org/matrix-android-sdk/blob/develop/CHANGES.rst)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

Signed-off-by: Courtney Cook <courtney1549@hotmail.com>

### CODE REVIEW

**Comment by bemica** [comment]
Signed-off-by: Nathan Beneke <nathan@nathanbeneke.com>

File: matrix-sdk/src/main/java/org/matrix/androidsdk/MXDataHandler.java
Comment (CMenne) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/MXDataHandler.java:2110]: This is a good change as it greatly reduces the message chains in Room.java. It really helps with readability.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java
Comment (CMenne) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java:2107]: Seems like a long method. Some of the if-else's could potentially be extracted into their own methods.

**Review by CMenne — CHANGES_REQUESTED** [review]
Overall, we think that the changes that were made in terms of message chains and the API callback in CreatedEvent helped the readability of the code. Unfortunately, the build is failing, so we can't approve the pull request until it passes. 

@Xaivteev @twarling @CMenne

**Review by nickplucker — APPROVED** [review]
I think the changes you made greatly improved the quality and readability of the code. The build seems to be failing for weird and not quite understood reasons that you didn't (probably) cause, so I'll approve this.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java
Comment (holma198) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java:2201]: Taking the callback out of this argument list is a good move.  I can't think of any value in having a new ApiCallback in that argument list be defined within it.  That being said, it would be nice to see `getCallback()` itself either get cut down further or moved off into a new file with `getLocalCB()` from [b2d51e90321c2d4f06a2f152727f14f5d42561a1] to put all these ApiCallback methods in one place.

**Review by holma198 — APPROVED** [review]
### A Summary of This Pull Request
________________________
#### Major Changes
-  retrieveRoomRestClient() - Though a simple change, removing a portion of the message chain removed some of the seemingly endless repetition in `Room.java`, this is definitely a good refactor
- getLocalCB() - the primary change appears to be a method call rather than an assignment to a new ApiCallback.  This, again, is a logical move.  The naming aids in making this complex set of ApiCallbacks make a bit more sense. 
- getCallback() - a more generalized version of the above mention, however, addresses messier code.  The change here was even more necessary as the new declaration was within a list of arguments which should be avoided.
#### Overall Thoughts
_____
 - The changes made were of great help for understanding what is happening in `Room.java`, though I believe much more work needs to be done (https://github.com/nathan-beneke/matrix-android-sdk/pull/5/commits/8c92e676d94136c8b96a68e3b7c7a9403e9370fc#r241645910), this is a good start on what appears to be a long refactoring process
- Lastly there seems to be some confusion on the Travis CI builds.  While it is true that Build 11 was failing, build 12 was passing, which was the last Travis run.  Build 11 specifically failed on `The command "./tools/travis/check_pr.sh" exited with 1.`, however this same line is resolved in Build 12.  With that said, I have no hesitation in approving this PR.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java
Comment (globbycodes) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java:2105]: I feel like the extraction here helps a bit with readability. I think I see why you did not extract it into a different class, I assume the security is the reason for that. But still, this extraction into a function helps.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/MXDataHandler.java
Comment (globbycodes) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/MXDataHandler.java:2110]: This is a nice way of reducing this chain. Especially because it is used multiple times in the code.

**Review by globbycodes — APPROVED** [review]
I think the changes are good overall. Definitely helps with the readability. So good job!

**Review by leonidscott — APPROVED** [review]
**Issues:** Github lets you link to a specific line of code. I think it would have helped with clarity if you had added those to your issues. I spent a fairly long time searching through directories in your project trying to find `Room.java`. Either the links, or just including the path to that file would have helped a lot. However, I was happy that you tagged issues in your commits.  

### matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java: 

**Regarding to Issue [#1]:** Extracting that function out is indeed slightly better then what was there before. The original block of code was super weird, I didn't even know you could do that in Java. Considering that it took me a long time to figure out what was going on was decent enough justification to move it into its own function. However, the uggly logic is still there. I suppose the reason you didn't create a separate class was for possible security reasons. Did you ever contact maintainers to see how you could have broken that object out into its own class or an abstract class?  

**Regarding to Issue [#2]:** Clearly, what was there before was not good. I'm just skeptical that this isn't much better. I suppose since it is a uncommonly large method it kind of makes sense, but now you have to worry about naming the new function, and hiding the problem.  Is there no common patterns between the code you changed in [#1] and this?  

**Regarding to Issue [#3]:**  Seems like a good change. However, there were no commits that tagged [#3].

**Regarding to Issue [#4]:** It's good that you commented on the issue and said that you were not going to peruse making those changes.

**Comment by NicMcPhee** [comment]
It's not entirely clear what to do about the half-passing build. We looked at it for a while last week, and it's just not obvious why it's failing. It's possible that simply merging it into master will fix the broken build, but it might not.

My sense is that to properly fix this would require communication with the Matrix team, and I don't want that to tie up the world. @CMenne – could you go ahead and OK this so they can merge into master locally and we can see what happens there? Thanks.

**Review by CMenne — APPROVED** [review]

**Comment by NicMcPhee** [comment]
Thank to @CMenne for approving this!

**Pull request merged**
